### PR TITLE
many: use symlinks for active component

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Use symlinks to point to the currently active component for a given snap revision, instead of mounting new component revisions always in the same mountpoint. This makes updates of components more atomic and will simplify parts of the code.

The symlinks will take the form

/snap/<snap_name>/components/<snap_revision>/<comp_name>

while the components will be mounted on

/snap/<snap_name>/components/mnt/<comp_name>/<comp_revision>

### CODE REVIEW

File: snap/component.go [hunk 1]
```diff
@@ -110,8 +110,8 @@ func (c *componentPlaceInfo) Filename() string {
 // will be of the form:
 // /snaps/<snap_instance>/components/<snap_revision>/<component_name>
```
Comment (bboozzoo) [inline snap/component.go:111]: the doc comment needs an update

File: overlord/snapstate/backend/link.go [hunk 2]
```diff
@@ -184,6 +185,25 @@ func (b Backend) LinkSnap(info *snap.Info, dev snap.Device, linkCtx LinkContext,
...
+	revStart := strings.LastIndex(mountDir, "_")
+	if revStart == -1 {
+		return "", fmt.Errorf("internal error: no revision in %q", mountDir)
+	}
+	return mountDir[0:revStart], nil
+}
+
+func (b Backend) LinkComponent(cpi snap.ContainerPlaceInfo) error {
+	mountDir := cpi.MountDir()
+	linkPath, err := componentLinkPath(mountDir)
```
Comment (bboozzoo) [inline overlord/snapstate/backend/link.go:199]: hmm, what if we extend snap.ContainerPlaceInfo to add eg. `ActiveSymlinkName()` which would return `/snap/<name>/current` for snaps and `/snap/<name>/components/<comp>` for components?

File: overlord/snapstate/backend/link.go [hunk 3]
```diff
@@ -184,6 +185,25 @@ func (b Backend) LinkSnap(info *snap.Info, dev snap.Device, linkCtx LinkContext,
 	return rebootInfo, nil
 }
 
+func componentLinkPath(mountDir string) (string, error) {
+	// The symlink to current is the mount with the stripped revision
+	revStart := strings.LastIndex(mountDir, "_")
+	if revStart == -1 {
+		return "", fmt.Errorf("internal error: no revision in %q", mountDir)
+	}
+	return mountDir[0:revStart], nil
```
Comment (bboozzoo) [inline overlord/snapstate/backend/link.go:194]: were `/snap/<name>/components/<comp>/current` and `/snap/<name>/components/<comp>/<rev>` maybe considered as paths we could use?

**Review by bboozzoo — COMMENTED** [review]
Some questions

File: overlord/snapstate/backend/link.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/link.go:194]: Yes, I was a bit between the two options, what I implemented is shorter and matches the path defined in the spec, and seems to me cleaner, but using `.../current` makes this consistent with snaps.

File: tests/main/component/task.yaml [hunk 4]
```diff
@@ -15,23 +15,38 @@ execute: |
...
+      readlink "$symlink" | MATCH "comp1_$comp_rev"
+      # and is seen from snap app
+      snap-with-comps.test
+  }
+
+  # Install, then reinstall local component
+  install_comp x1
+  install_comp x2
+
+  # TODO: remove checks when implemented by snapd
```
Comment (zyga) [inline tests/main/component/task.yaml:52]: Which checks?

File: snap/component.go [hunk 5]
```diff
@@ -110,8 +110,8 @@ func (c *componentPlaceInfo) Filename() string {
...
 func (c *componentPlaceInfo) MountDir() string {
-	return filepath.Join(BaseDir(c.snapInstance), "components",
-		c.snapRevision.String(), c.compName)
+	return filepath.Join(BaseDir(c.snapInstance), "components", c.snapRevision.String(),
+		fmt.Sprintf("%s_%s", c.compName, c.compRevision.String()))
```
Comment (pedronis) [inline snap/component.go:114]: @alfonsosanchezbeato this kind of path doesn't exist in the spec and is a consequence of the change here, they don't look particularly great tbh. We need to discuss what options we have, using "current" for the final target is a deviation from the spec as is but also adds another level that is not very clear we need for components except for this intermediate paths for the mountpoints. We should discuss on this

File: tests/main/component/task.yaml (hunk 4 above)
Comment (alfonsosanchezbeato) [reply tests/main/component/task.yaml]: I've rephrased this, was not really talking about removing test checks :slightly_smiling_face:

File: overlord/snapstate/backend/link.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/link.go]: FTR the agreement has been:
* Symlinks paths will still be `/snap/<snap>/components/<snap_revision>/<component_name>`
* Components will be mounted on `/snap/<snap>/components/mnt/<comp_name>/<comp_revno>`

File: overlord/snapstate/backend/link.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/link.go]: I've been thinking about this, but I'm not fully convinced as strictly speaking `ContainerPlaceInfo` does not need to know what is an active snap/component or how those reference to the containers.

File: snap/component.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply snap/component.go]: Changed now

File: snap/component.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply snap/component.go]: Changed now as discussed

**Review by alfonsosanchezbeato — COMMENTED** [review]
After agreement on where the components will be mounted, this is ready for a new pass.

**Review by pedronis — APPROVED** [review]
thank you

File: overlord/snapstate/backend/link.go [hunk 6]
```diff
@@ -184,6 +185,35 @@ func (b Backend) LinkSnap(info *snap.Info, dev snap.Device, linkCtx LinkContext,
...
+	return filepath.Join(compBase, snapRev.String(), compName)
+}
+
+func (b Backend) LinkComponent(cpi snap.ContainerPlaceInfo, snapRev snap.Revision) error {
+	mountDir := cpi.MountDir()
+	linkPath := componentLinkPath(cpi, snapRev)
+
+	// We want a relative symlink, so we remove the common parts of the
+	// path and then go up the needed directories.
+	linkComps := strings.Split(linkPath, "/")
```
Comment (bboozzoo) [inline overlord/snapstate/backend/link.go:200]: would `filepath.Rel()` work here?

Comment (alfonsosanchezbeato) [reply overlord/snapstate/backend/link.go]: Right, it makes this easier, changed now.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: snap/component.go [hunk 7]
```diff
@@ -70,29 +70,33 @@ func (csi *ComponentSideInfo) Equal(other *ComponentSideInfo) bool {
 	return *csi == *other
 }
 
+// ComponentBaseDir returns where components are to be found for the
+// snap with name instanceName.
+func ComponentsBaseDir(instanceName string) string {
+	return filepath.Join(BaseDir(instanceName), "components")
```
Comment (zyga) [inline snap/component.go:76]: Is "components" a reserved snap name both in the store and in snapd itself?

File: tests/main/component/task.yaml [hunk 8]
```diff
@@ -15,23 +15,39 @@ execute: |
...
+      snap change "$chg_id" | MATCH "^Done .*Mount component"
+      snap change "$chg_id" | MATCH "^Done .*Make component .* available to the system"
+
+      # File has been copied around
+      comp_inst_path=/var/lib/snapd/snaps/snap-with-comps+comp1_${comp_rev}.comp
+      stat "$comp_inst_path"
+
+      # Component is mounted (note that we need to escape the "+" in the path)
+      # x1 is the snap revision
+      symlink=/snap/snap-with-comps/components/x1/comp1
```
Comment (zyga) [inline tests/main/component/task.yaml:39]: Can you extend this test to run on Fedora and show that components do not break there and that the locations do not depend on /snap being a symbolic link?

File: snap/component.go (hunk 7 above)
Comment (alfonsosanchezbeato) [reply snap/component.go:76]: I don't know, but it is under instance name so it won't conflict.

**Review by zyga — APPROVED** [review]
Woot, thank you!

**Pull request merged**